### PR TITLE
Refactor dora meta manager

### PR DESCRIPTION
**Pull request by elega**
### What changes are proposed in this pull request?

1. Move metastore into dora meta manager
2. Handles listing cache invalidation
3. Add unit tests
4. Introduced caffeinine cache

### Why are the changes needed?

Dora metadata management has a lot of consistency issue between metadata/listing/page cahces. To address these issues, the first step is to introduce a centralized management place and put these logic together.

### Does this PR introduce any user facing changes?

N/A

### CODE REVIEW

File: dora/core/common/src/main/java/alluxio/exception/ListOnNonDirectoryException.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+package alluxio.exception;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Exception for situations where a listing operation happens on a non-directory file/object.
+ */
+@ThreadSafe
+public class ListOnNonDirectoryException extends AlluxioException {
```
Comment (huanghua78) [inline dora/core/common/src/main/java/alluxio/exception/ListOnNonDirectoryException.java:20]: I have the doubt that we don't need such exception at all.

File: dora/core/server/worker/src/main/java/alluxio/worker/dora/DoraMetaManager.java [hunk 2]
```diff
@@ -35,25 +40,38 @@
...
   private final UnderFileSystem mUfs;
   private boolean mPopulateMetadataFingerprint =
       Configuration.getBoolean(PropertyKey.DORA_WORKER_POPULATE_METADATA_FINGERPRINT);
 
+  private final long mListingCacheCapacity
+      = Configuration.getInt(PropertyKey.DORA_UFS_LIST_STATUS_CACHE_NR_FILES);
+  private final Cache<String, ListStatusResult> mListStatusCache = mListingCacheCapacity == 0
+      ? null
+      : Caffeine.newBuilder()
+      .maximumWeight(Configuration.getInt(PropertyKey.DORA_UFS_LIST_STATUS_CACHE_NR_FILES))
```
Comment (huanghua78) [inline dora/core/server/worker/src/main/java/alluxio/worker/dora/DoraMetaManager.java:56]: s/Configuration.getInt(PropertyKey.DORA_UFS_LIST_STATUS_CACHE_NR_FILES)/mListingCacheCapacity

File: dora/core/server/worker/src/main/java/alluxio/worker/dora/DoraMetaManager.java [hunk 3]
```diff
@@ -142,24 +161,157 @@ public void put(String path, FileStatus meta) {
...
+   * otherwise an option contains an ufs status array.
+   * @throws IOException if the UFS call failed
+   */
+  public Optional<UfsStatus[]> listFromUfsThenCache(String path, boolean isRecursive)
+      throws IOException {
+    if (mListStatusCache == null) {
+      return listFromUfs(path, isRecursive);
+    }
+    try {
+      var cached = mListStatusCache.get(path, (k) -> {
```
Comment (huanghua78) [inline dora/core/server/worker/src/main/java/alluxio/worker/dora/DoraMetaManager.java:229]: `var` is a new syntax introduced by JDK11?
Can you avoid this? Or you prefer this way?

File: dora/core/server/worker/src/main/java/alluxio/worker/dora/DoraMetaManager.java [hunk 4]
```diff
@@ -142,24 +161,157 @@ public void put(String path, FileStatus meta) {
...
+        try {
+          Optional<UfsStatus[]> listResults = listFromUfs(path, isRecursive);
+          return listResults.map(
+                  ufsStatuses -> new ListStatusResult(
+                      System.nanoTime(), ufsStatuses,
+                      ufsStatuses.length == 1 && ufsStatuses[0].isFile()
+                  ))
+              .orElseGet(() -> new ListStatusResult(System.nanoTime(), null, false));
```
Comment (huanghua78) [inline dora/core/server/worker/src/main/java/alluxio/worker/dora/DoraMetaManager.java:237]: It's better to add comment here to explain "negative result"?
Or adding comments in front of the function.

**Comment by huanghua78** [comment]
Looks good. But please ask @ARB to review the new dependent package.

File: dora/core/server/worker/src/main/java/alluxio/worker/dora/DoraMetaManager.java (hunk 2 above)
Comment (elega) [reply dora/core/server/worker/src/main/java/alluxio/worker/dora/DoraMetaManager.java:56]: ack

File: dora/core/server/worker/src/main/java/alluxio/worker/dora/DoraMetaManager.java (hunk 3 above)
Comment (elega) [reply dora/core/server/worker/src/main/java/alluxio/worker/dora/DoraMetaManager.java:229]: yeah i switched it back

File: dora/core/server/worker/src/main/java/alluxio/worker/dora/DoraMetaManager.java (hunk 4 above)
Comment (elega) [reply dora/core/server/worker/src/main/java/alluxio/worker/dora/DoraMetaManager.java:221]: done

Comment (huanghua78) [reply dora/core/server/worker/src/main/java/alluxio/worker/dora/DoraMetaManager.java:221]: OK

**Review by huanghua78 — APPROVED** [review]

**Comment by elega** [comment]
alluxio-bot, merge this please.

**Pull request merged**